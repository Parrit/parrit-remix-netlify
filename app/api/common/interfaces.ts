export interface LoginRequest {
  projectName?: string;
  password?: string;
}

export interface LoginResult {
  message: string;
}

export interface Project {
  id: string;
  name: string;
  pairingBoards: PairingBoard[];
  people: Person[];
}

export interface PairingBoard {
  id: string;
  name: string;
  exempt: boolean;
  people: Person[];
  roles: Role[];
}

export interface Person {
  id: string;
  name: string;
}

export interface Role {
  id: string;
  name: string;
}

export interface PairingArrangementDTO {
  id: number; // this should be changed to string
  pairingTime: string;
  pairingHistories: PairingHistory[];
}

export interface PairingHistory {
  pairingBoardName: string;
  people: Person[];
}
