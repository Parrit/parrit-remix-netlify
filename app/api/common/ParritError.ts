export interface IParritError<T> {
  server?: string;
  fields?: Record<keyof T, string | null>;
}

export class ParritError<T> extends Error {
  public readonly data: IParritError<T>;

  constructor(arg: IParritError<T>) {
    super(JSON.stringify(arg));
    this.data = arg;
  }

  toResponse(status = 400): Response {
    return new Response(JSON.stringify(this.data), { status });
  }

  static obscured() {
    return new ParritError({ server: "Server error" });
  }

  static fromString<T>(str?: string) {
    if (!str) {
      return undefined;
    }
    const obj = JSON.parse(str) as IParritError<T>;
    if (this.isParritError<T>) {
      return new ParritError(obj);
    }
    return undefined;
  }

  private static isParritError<T>(obj: object): obj is ParritError<T> {
    const cast = obj as ParritError<T>;
    return !!cast.data.server || !!cast.data.fields;
  }
}
