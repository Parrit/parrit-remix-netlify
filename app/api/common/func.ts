import {
  FLOATING_IDX,
  PairingBoard,
  Person,
  Project,
} from "./interfaces/parrit.interfaces";
import * as R from "ramda";

// Functional programing utilities for working with interfaces

export const movePerson = (
  project: Project,
  person: Person,
  position: PairingBoard
) =>
  R.pipe<[Project], Project, Project>(
    removePerson(person),
    addPerson(person, position)
  )(project);

export const addPerson = R.curry(
  (person: Person, position: PairingBoard, project: Project) => {
    const f1 = addPersonToPosition(person, position);
    const f2 =
      position.id === FLOATING_IDX ? addPersonToFloating(person) : R.identity;
    return R.pipe<[Project], Project, Project>(f1, f2)(project);
  }
);

const addPersonToPosition = R.curry(
  (person: Person, position: PairingBoard, project: Project) =>
    R.over<Project, PairingBoard[]>(
      R.lensProp("pairingBoards"),
      R.map<PairingBoard, PairingBoard>((x) => {
        if (x.id === position.id) {
          return R.over<PairingBoard, Person[]>(
            R.lensProp("people"),
            R.append({ ...person, pairing_board_id: position.id }),
            x
          );
        }
        return x;
      }),
      project
    )
);

const addPersonToFloating = R.curry((person: Person, project: Project) =>
  R.over<Project, Person[]>(
    R.lensPath(["floating", "people"]),
    R.append(person),
    project
  )
);

export const removePerson = R.curry(
  (person: Person, project: Project): Project =>
    R.pipe<[Project], Project, Project>(
      removePersonFromBoards(person),
      removePersonFromFloating(person)
    )(project)
);

const removePersonFromBoards = R.curry(
  (person: Person, project: Project): Project =>
    R.over<Project, PairingBoard[]>(
      R.lensProp("pairingBoards"),
      R.map<PairingBoard, PairingBoard>((x) => {
        if (x.people.includes(person)) {
          return R.set<PairingBoard, Person[]>(
            R.lensProp("people"),
            R.without(x.people, [person]),
            x
          );
        }
        return x;
      }),
      project
    )
);

const removePersonFromFloating = R.curry(
  (person: Person, project: Project): Project =>
    R.set<Project, Person[]>(
      R.lensPath(["floating", "people"]),
      R.without([person], project.floating.people),
      project
    )
);

const pbIndexById = R.curry((pbs: PairingBoard[], pb_id: string) =>
  R.findIndex<PairingBoard>((pb) => pb.id === pb_id)(pbs)
);

const renamePairingBoardFromList = R.curry(
  (pb_id: string, rename: string, pbs: PairingBoard[]) => {
    const foundIndex = pbIndexById(pbs, pb_id);
    if (foundIndex < 0) {
      return pbs;
    }
    return R.adjust<PairingBoard>(foundIndex, R.assoc("name", rename), pbs);
  }
);

export const renamePairingBoard = R.curry(
  (project: Project, pb_id: string, rename: string): Project =>
    R.over<Project, PairingBoard[]>(
      R.lensProp("pairingBoards"),
      renamePairingBoardFromList(pb_id, rename),
      project
    )
);

export const canAPairingBeMade = (project: Project): boolean => {
  const atLeast2Floaters = project.floating.people.length >= 2;
  const atLeast1Floater = project.floating.people.length >= 1;
  const atLeast1EmptyBoard = getEmptyPairingBoard(project) !== undefined;
  const atLeast1UnpairedSticker = unpairedStickingPeople(project).length >= 1;
  return (
    (atLeast1Floater && atLeast1UnpairedSticker) ||
    (atLeast2Floaters && atLeast1EmptyBoard)
  );
};

export const getEmptyPairingBoard = (
  project: Project
): PairingBoard | undefined => {
  return project.pairingBoards.find(
    (pb) => pb.people.length === 0 && !pb.exempt
  );
};

const unpairedStickingPeople = (project: Project): Person[] =>
  project.pairingBoards
    .filter((board) => !board.exempt)
    .filter((board) => board.people.length === 1)
    .flatMap((board) => board.people);

const findPairingBoardByPerson = (
  project: Project,
  person: Person
): PairingBoard | undefined =>
  project.pairingBoards.find((pb) => pb.people.find((p) => p.id === person.id));
