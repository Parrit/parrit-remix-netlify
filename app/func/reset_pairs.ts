import {
  FLOATING_IDX,
  Project,
} from "~/api/common/interfaces/parrit.interfaces";

export default (project: Project) => {
  const boards = project.pairingBoards;

  const people = project.people.map((p) => {
    const board =
      boards.find((b) => b.id === p.pairing_board_id) ?? project.floating;
    if (board.exempt || board.id === FLOATING_IDX) {
      return p;
    }
    return { ...p, pairing_board_id: FLOATING_IDX };
  });

  return { ...project, people };
};
