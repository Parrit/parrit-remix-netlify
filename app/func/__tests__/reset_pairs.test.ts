import reset_pairs from "../reset_pairs";
import { PROJECT_START } from "./common";
import { describe, expect, it } from '@jest/globals';

describe("reset_pairs", () => {
  it("moves everything from non-exempt PBs to floating", () => {
    const result = reset_pairs(PROJECT_START);
    expect(result).toMatchSnapshot();
  });
});
