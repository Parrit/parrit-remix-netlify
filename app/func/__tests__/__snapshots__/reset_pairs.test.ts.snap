// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reset_pairs moves everything from non-exempt PBs to floating 1`] = `
{
  "floating": {
    "exempt": false,
    "id": "-1",
    "name": "floating_board_name",
  },
  "id": "project_id",
  "name": "project_name",
  "pairingBoards": [
    {
      "exempt": false,
      "id": "red_id",
      "name": "red_start",
    },
    {
      "exempt": false,
      "id": "blue_id",
      "name": "blue_start",
    },
    {
      "exempt": true,
      "id": "exempt",
      "name": "out_of_office",
    },
  ],
  "people": [
    {
      "id": "red_parrit_id",
      "name": "red_parrit",
      "pairing_board_id": "-1",
      "type": "Person",
    },
    {
      "id": "absent",
      "name": "Permanent Vacation",
      "pairing_board_id": "exempt",
      "type": "Person",
    },
    {
      "id": "floater_id",
      "name": "floater_name_start",
      "pairing_board_id": "-1",
      "type": "Person",
    },
  ],
  "roles": [
    {
      "id": "blue_role",
      "name": "Blue Role",
      "pairing_board_id": "blue_id",
      "type": "Role",
    },
    {
      "id": "ooo",
      "name": "OOOROLE",
      "pairing_board_id": "exempt",
      "type": "Role",
    },
  ],
}
`;
